flexdll.c

```diff
@@ -496,8 +498,22 @@ void *flexdll_wdlopen(const wchar_t *file, int mode) {
 #endif /* __STDC_SECURE_LIB__ >= 200411L*/
 #endif /* CYGWIN */
 
+again:
+  if (units_mutex == INVALID_HANDLE_VALUE) {
+    HANDLE hMutex = CreateMutex(NULL, TRUE, NULL);
+    if (InterlockedCompareExchangePointer(&units_mutex, hMutex, INVALID_HANDLE_VALUE) != INVALID_HANDLE_VALUE) {
+      CloseHandle(hMutex);
+      goto again;
+    }
+  } else {
+    if (WaitForSingleObject(units_mutex, INFINITE) == WAIT_FAILED) {
+      if (!err->code) err->code = 1;
+      return NULL;
+    }
+  }
```

Comment:
I find the `goto again` a bit convoluted, isn't a simple while loop equivalent?

```suggestion
  while (units_mutex == INVALID_HANDLE_VALUE) {
    HANDLE hMutex = CreateMutex(NULL, TRUE, NULL);
    if (hMutex == NULL) /* do something */;
    if (InterlockedCompareExchangePointer(&units_mutex, hMutex, INVALID_HANDLE_VALUE) != INVALID_HANDLE_VALUE)
      CloseHandle(hMutex);
    else break; /* optional */
  }
  
  if (WaitForSingleObject(units_mutex, INFINITE) == WAIT_FAILED) {
    if (!err->code) err->code = 1;
      return NULL;
    }
  }
```

Reply:
I would think so. However I must either (a) be missing something or (b) messing up, because with that change `tests/lib-dynlink-domains` seems to loop infinitely :grimacing:

Reply:
I tried this suggestion afresh this morning - and still encounter a loop (and eventually a timeout).
Are you experiencing the same @MisterDA?

Reply:
After some off-line, collective head scratching with @MisterDA (who was also observing looping), we've established that the while-loop is not equivalent, as it would attempt a `WaitForSingleObject` after a successful `CreateMutex` and `InterlockedCompareExchangePointer` whereas the label-version doesn't.